draft/jumpstarter.rst

```diff
@@ -64,6 +64,115 @@ the other features of Python's ``async/await`` programming model.
+* Stopped --> The state of the actor after it has finished all of its ``stopping`` activities (think about how when you power off a computer).
+* Crashed --> The state of the actor when an exception was raised during startup or shutdown.
+
+Within those parent states, we have sub-states. For example:
+
+* Starting
+  * Dependencies Started --> The state of the actor after all of the actor's dependencies have been started.
+  * Resources Acquired --> The state of the actor after all resources have been acquired.
+  * Tasks Started --> The state of the actor after all tasks have been started.
+* Started
+  * Paused --> The state of the actor when all tasks are halted without shutting down the entire actor.
+  * Running --> The state of the actor when all tasks are running.
+    * Healthy --> The state of the actor when the actor is functioning properly.
+    * Degraded --> The state of the actor when the actor is not functioning properly but is still able to perform some of its duties.
+    * Unhealthy --> The state of the actor when the actor is temporarily not functioning.
+* Stopping
+  * Tasks Stopped --> The state of the actor after all tasks have been started.
+  * Resources Released --> The state of the actor after all resources have been acquired.
+  * Dependencies Stopped --> The state of the actor after all of the actor's dependencies have been started.
+
+In order to effectively model these states in Python, we propose using the mature `transitions`_ library, along with the `transitions-anyio`_ library. This gives us:
+
+1. Mature Hierarchical State Machine library support thanks to `transitions`_.
+2. Asynchronous state machine transitions (opening up abilities for concurrency, parallelization, and the latest ``async/await`` python support that's part of the motivation of this CEP in the first place) with `AnyIO`_ (thanks to `transitions-anyio`_) to abstract away the specific event loop of choice (like `AsyncIO`_, `Trio`_, or potentially others in the future).
+3. Native support within `transitions`_ for integrating with ``diagrams``/``graphviz`` to generate state machine diagrams (like the one below). Additionally, `transitions-gui`_ provides some interesting and promising capabilities for future Celery Flower-like projects to be able to visualize in a live, animated fashion the various Jumpstarter Actors and their states as transitions happen across all the various actors within the system.
+
+For a high level view, the parent states, their substates, and the transitions between them can be seen in the diagram below:
+
+TODO: Insert Jumpstarter State Machine Diagram Here: https://user-images.githubusercontent.com/48936/107506089-43225a00-6ba6-11eb-810e-0ac14bf0e1e9.png
+
+Also, in that diagram you can also see the ``Restart`` state. We propose a separate state machine which we'll call *Actor Restart State Machine* that models the Actor's state as it relates to restarts:
```

Comment:
This should be a parallel state in our state machine but transitions doesn't support arbitrary parallel states, only initial parallel states.
This is an open item in our implementation to resolve.
Currently, the draft implementation nests state machines in the parent state machine and we model transitions using conditions on the transition itself.